modules/ocf_www/files/build-vhosts

```diff
@@ -14,190 +11,350 @@ from pathlib import Path
             vhosts,
-            key=lambda vhost: (vhost.user, vhost.fqdn, vhost.port),
+            key=lambda vhost: (vhost.user, vhost.fqdn, bool(vhost.ssl)),
         )
-    ) + '\n'
+    )
+
 
+def test_and_overwrite_config(config_path, new_config):
+    """Diffs and tests the new config and overwrites the old config if
+    the test passes.
+
+    Returns whether the old config was overwritten.
+    """
+    (_, new_path), (__, old_path) = (
+        tempfile.mkstemp(
+            # Prefix with '.' so temp files don't get included
+            prefix='.' + os.path.basename(config_path),
+            dir=os.path.dirname(config_path),
+        )
+        for i in range(2)
+    )
+    try:
+        os.chmod(new_path, 0o644)
+        with open(new_path, 'w') as f:
+            f.write(new_config)
+        report('Wrote to ' + new_path)
 
-def main(argv=None):
+        report('Diff between current and new file:')
+        ret = subprocess.call(('diff', '-u', config_path, new_path))
+        report('[end diff]')
+
+        if not ret:
+            report('Nothing changed, not doing anything.')
+            return False
+
+        # Save existing config
+        shutil.copyfile(config_path, old_path)
+        report('Copied existing config to ' + old_path)
+
+        # Overwrite with new config and test
+        os.rename(new_path, config_path)
+
+        report('Performing config test.')
+        if target == 'web':
+            ret = subprocess.call(('apachectl', 'configtest'))
+        else:
+            # nginx is moronic and prints "test successful" on stderr,
+            # so echo to stderr only if nginx -t failed
+            p = subprocess.Popen(('nginx', '-t'), stderr=subprocess.PIPE)
+            ret = p.wait()
+            if ret:
+                report(p.stderr.read().decode('utf-8'), file=sys.stderr)
+
+        if ret:
+            report('Test failed! Replacing with old config.')
+            os.rename(old_path, config_path)
+
+        return not ret
+    finally:
+        for p in old_path, new_path:
+            try:
+                os.remove(p)
+            except FileNotFoundError:
+                pass
+
+
+def process_app_vhosts():
+    """Perform extra tasks specific to app vhosts.
+
+    This includes checking for membership in the ocfdev group, fixing
+    socket permissions, and generating SSL bundles."""
+    def groups_for_user(user):
+        return (g.gr_name for g in grp.getgrall() if user in g.gr_mem)
+
+    for domain, vhost in get_app_vhosts().items():
+        user = vhost['username']
+
+        # Warn about ocfdev group
+        if 'ocfdev' not in groups_for_user(user):
+            report(
+                "Warning: user '{}' not in group ocfdev but has app vhost '{}'"
+                .format(user, domain)
+            )
+
+        # Fix socket dir
+        socket_dir = str(APP_DIR / user)
+        try:
+            os.makedirs(socket_dir, 0o755)
+        except FileExistsError:
+            pass
```

Comment:
i didn't actually realize that existed for `os.makedirs` -- but since it does, that's actually probably a good idea

(my comment was not at all sarcastic, i was commenting on how this _was_ the correct way to do that. i didn't realize python 3.2 added this option)